dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/trackedentity/JdbcTrackedEntityChangeLogStore.java

```diff
@@ -0,0 +1,112 @@
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.tracker.export.trackedentity;
+
+import java.util.List;
+import java.util.Set;
+import lombok.RequiredArgsConstructor;
+import org.hisp.dhis.program.UserInfoSnapshot;
+import org.hisp.dhis.tracker.export.trackedentity.TrackedEntityChangeLog.Change;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.jdbc.core.namedparam.SqlParameterSource;
+import org.springframework.stereotype.Repository;
+
+@Repository("org.hisp.dhis.tracker.export.trackedentity.JdbcTrackedEntityChangeLogStore")
+@RequiredArgsConstructor
+public class JdbcTrackedEntityChangeLogStore {
+
+  private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;
+
+  private static final RowMapper<TrackedEntityChangeLog> customTrackedEntityChangeLogRowMapper =
+      (rs, rowNum) -> {
+        UserInfoSnapshot createdBy = new UserInfoSnapshot();
+        createdBy.setUsername(rs.getString("userName"));
+        createdBy.setFirstName(rs.getString("firstname"));
+        createdBy.setSurname(rs.getString("surname"));
+        createdBy.setUid(rs.getString("useruid"));
+
+        return new TrackedEntityChangeLog(
+            createdBy,
+            rs.getTimestamp("updatedAt"),
+            rs.getString("type"),
+            new Change(
+                new TrackedEntityChangeLog.TrackedEntityAttributeChange(
+                    rs.getString("trackedEntityAttributeUid"),
+                    rs.getString("previousValue"),
+                    rs.getString("currentValue"))));
+      };
+
+  public List<TrackedEntityChangeLog> getTrackedEntityChangeLog(
+      String trackedEntity, Set<String> attributes) {
+    String sql =
+        """
+          select cl.audittype,
+              case
+          when cl.audittype = 'CREATE' then cl.currentchangelogvalue
+          when cl.audittype = 'UPDATE' then cl.currentchangelogvalue
+          end as currentValue,
+              case
+          when cl.audittype = 'DELETE' then cl.currentchangelogvalue
+          when cl.audittype = 'UPDATE' then cl.previouschangelogvalue
+          end as previousValue, cl.created as updatedAt, cl.modifiedby as userName, cl.trackedEntityAttributeUid, cl.firstname, cl.surname, cl.username, cl.useruid, cl.audittype as type
```

Comment:
I would only add aliases in one place in the query. That makes it easier to follow. I moved all aliases into the subquery https://github.com/dhis2/dhis2-core/pull/16622/files

You also have a duplicate with different casing that is confusing `cl.username` and `cl.modifiedby as userName`. Store in https://github.com/dhis2/dhis2-core/pull/16622/files addresses these.